### PR TITLE
feat: configurable app board

### PR DESCRIPTION
**Pull request by hanszhoou**
## Description

Closes # (NO_ISSUE)
https://mask.atlassian.net/browse/MF-138

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/plugins/Wallet/src/messages.ts [hunk 1]
```diff
@@ -237,3 +239,8 @@ if (import.meta.webpackHot) import.meta.webpackHot.accept()
 export const WalletMessages: { events: PluginMessageEmitter<WalletMessage> } = {
     events: createPluginMessage<WalletMessage>(PLUGIN_ID),
 }
+
+export const MaskMessages = new WebExtensionMessage<MaskEvents>({ domain: 'mask' })
```
Comment (Jack-Works) [inline packages/plugins/Wallet/src/messages.ts]: Please don't move it here. `MaskMessages` is for general purposes.

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/plugins/Wallet/src/messages.ts (hunk 1 above)
Comment (hanszhoou) [reply packages/plugins/Wallet/src/messages.ts]: How to use `MaskMessages` in `packages/plugins/FileServices`?

Comment (Jack-Works) [reply packages/plugins/Wallet/src/messages.ts]: please hold on. I need to consider how to handle this

Comment (Jack-Works) [reply packages/plugins/Wallet/src/messages.ts]: @zhouhanseng https://github.com/DimensionDev/Maskbook/pull/5961 as a temporary solution.

File: packages/mask/background/database/avatar-cache/avatar.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import { ProfileIdentifier } from '@masknet/shared-base'
 import { queryAvatarDB, isAvatarOutdatedDB, storeAvatarDB, IdentifierWithAvatar, createAvatarDBAccess } from './db'
-import { MaskMessages } from '../../../shared'
+import { MaskMessages } from '@masknet/plugin-wallet'
```
Comment (guanbinrui) [inline packages/mask/background/database/avatar-cache/avatar.ts]: Move this migration to another PR.

Comment (Jack-Works) [reply packages/mask/background/database/avatar-cache/avatar.ts]: don't do this migration. it should not be moved to '@masknet/plugin-wallet'

**Comment by hanszhoou** [comment]
@DimensionDev/qa

File: packages/mask/src/plugins/RedPacket/SNSAdaptor/index.tsx [hunk 3]
```diff
@@ -20,6 +20,8 @@ import RedPacketDialog from './RedPacketDialog'
 import { RedPacketInPost } from './RedPacketInPost'
 import { RedPacketNftInPost } from './RedPacketNftInPost'
 import { RedPacketIcon, NFTRedPacketIcon } from '@masknet/icons'
+import { ApplicationEntry } from '@masknet/shared'
+import { requestComposition } from '@masknet/shared-base'
```
Comment (Jack-Works) [inline packages/mask/src/plugins/RedPacket/SNSAdaptor/index.tsx]: `requreComposition` should not be an exported function in `'@masknet/shared-base'`.

It should be provided by the first argument of `init()`, and the plugin should not provide the plugin ID on it's own.

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/mask/src/plugins/RedPacket/SNSAdaptor/index.tsx (hunk 3 above)
Comment (Jack-Works) [reply packages/mask/src/plugins/RedPacket/SNSAdaptor/index.tsx]: Please add it in plugin context for every plugin.

Comment (hanszhoou) [reply packages/mask/src/plugins/RedPacket/SNSAdaptor/index.tsx]: done

File: packages/plugin-infra/src/types.ts [hunk 4]
```diff
@@ -124,6 +124,7 @@ export namespace Plugin.Shared {
         contribution?: Contribution
         /** Declare ability this plugin supported. */
         ability?: Ability
+        /** Declare relative application entries of the plugin to render */
```
Comment (guanbinrui) [inline packages/plugin-infra/src/types.ts]: Does this comment relate to anything?

File: packages/mask/src/plugins/Trader/SNSAdaptor/index.tsx [hunk 5]
```diff
@@ -18,6 +22,29 @@ const sns: Plugin.SNSAdaptor.Definition = {
         )
     },
     enhanceTag,
+    ApplicationEntries: [
+        {
+            RenderEntryComponent(key) {
+                const currentPluginId = usePluginIDContext()
+                const chainId = useChainId()
+                const isHidden =
+                    currentPluginId !== NetworkPluginID.PLUGIN_EVM ||
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/index.tsx]: Can these condition statements move to the place where we render entries?

File: packages/mask/src/plugins/Savings/SNSAdaptor/index.tsx [hunk 6]
```diff
@@ -1,9 +1,36 @@
...
 const sns: Plugin.SNSAdaptor.Definition = {
     ...base,
     init(signal) {},
+    ApplicationEntries: [
+        {
+            RenderEntryComponent(key) {
+                const currentPluginId = usePluginIDContext()
+                const chainId = useChainId()
+                const [open, setOpen] = useState(false)
+                const isHidden =
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Savings/SNSAdaptor/index.tsx]: Can these condition statements move to the place where we render entries?

File: packages/mask/src/plugins/Pets/SNSAdaptor/index.tsx [hunk 7]
```diff
@@ -14,6 +20,33 @@ const sns: Plugin.SNSAdaptor.Definition = {
...
         )
     },
+    ApplicationEntries: [
+        {
+            RenderEntryComponent(key) {
+                const currentPluginId = usePluginIDContext()
+                const chainId = useChainId()
+                const currentSNSNetwork = getCurrentSNSNetwork(activatedSocialNetworkUI.networkIdentifier)
+
+                const isHidden =
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Pets/SNSAdaptor/index.tsx]: Can these condition statements move to the place where we render entries?

File: packages/mask/src/plugins/MaskBox/SNSAdaptor/index.tsx [hunk 8]
```diff
@@ -26,6 +27,36 @@ const sns: Plugin.SNSAdaptor.Definition = {
...
     },
+    ApplicationEntries: [
+        {
+            RenderEntryComponent(key) {
+                return (
+                    <div key={key}>
+                        <ApplicationEntry
+                            title="Mask Bridge"
+                            icon={new URL('../assets/bridge.png', import.meta.url).toString()}
+                            onClick={() => window.open('https://bridge.mask.io/#/', '_blank', 'noopener noreferrer')}
```
Comment (guanbinrui) [inline packages/mask/src/plugins/MaskBox/SNSAdaptor/index.tsx]:
```ts
import { openWindow } from '@masknet/shared-base-ui'
```

File: packages/mask/src/plugins/MaskBox/SNSAdaptor/index.tsx [hunk 9]
```diff
@@ -26,6 +27,36 @@ const sns: Plugin.SNSAdaptor.Definition = {
...
+            defaultSortingPriority: 5,
+        },
+        {
+            RenderEntryComponent(key) {
+                return (
+                    <div key={key}>
+                        <ApplicationEntry
+                            title="MaskBox"
+                            icon={new URL('../assets/mask_box.png', import.meta.url).toString()}
+                            onClick={() => window.open('https://box.mask.io/#/', '_blank', 'noopener noreferrer')}
```
Comment (guanbinrui) [inline packages/mask/src/plugins/MaskBox/SNSAdaptor/index.tsx]:
```ts
import { openWindow } from '@masknet/shared-base-ui'
```

File: packages/mask/src/plugins/ITO/SNSAdaptor/index.tsx [hunk 10]
```diff
@@ -57,6 +62,56 @@ const sns: Plugin.SNSAdaptor.Definition = {
             ),
         },
     },
+    ApplicationEntries: [
+        {
+            RenderEntryComponent(key) {
+                const currentPluginId = usePluginIDContext()
+                const chainId = useChainId()
+                const isHidden =
```
Comment (guanbinrui) [inline packages/mask/src/plugins/ITO/SNSAdaptor/index.tsx]: Can these condition statements move to the place where we render entries?

File: packages/mask/src/plugins/FindTruman/SNSAdaptor/index.tsx [hunk 11]
```diff
@@ -84,6 +92,25 @@ const sns: Plugin.SNSAdaptor.Definition = {
         if (!link) return null
         return <Renderer url={link} />
     },
+    ApplicationEntries: [
+        {
+            RenderEntryComponent(key) {
+                const [open, setOpen] = useState(false)
+                const currentPluginId = usePluginIDContext()
```
Comment (Jack-Works) [inline packages/mask/src/plugins/FindTruman/SNSAdaptor/index.tsx]: This hook name is confusing. Maybe `useCurrentWeb3NetworkPluginID()`?

File: packages/mask/src/components/shared/ApplicationBoard.tsx [hunk 12]
```diff
@@ -86,350 +20,23 @@ const useStyles = makeStyles()((theme) => {
...
-                            </div>
-                        ) : null,
-                )}
+                {snsAdaptorPlugins
+                    .reduce<Plugin.SNSAdaptor.ApplicationEntry[]>((acc, cur) => {
+                        if (!cur.ApplicationEntries) return acc
+                        return acc.concat(cur.ApplicationEntries ?? [])
+                    }, [])
+                    .sort((a, b) => a.defaultSortingPriority - b.defaultSortingPriority)
+                    .map((x, i) => x.RenderEntryComponent(i))}
```
Comment (Jack-Works) [inline packages/mask/src/components/shared/ApplicationBoard.tsx]: Please do not let the plugin handle the `key` property and do not call components as function (this will fragile when those components use hooks). Do this:

```js
.map(X => <React.Fragment key={X.ID}><X.RenderEntryComponent /></React.Fragment>)
```

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/mask/src/plugins/FindTruman/SNSAdaptor/index.tsx (hunk 11 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/FindTruman/SNSAdaptor/index.tsx]: I think so.

File: packages/mask/src/components/shared/ApplicationBoard.tsx (hunk 12 above)
Comment (hanszhoou) [reply packages/mask/src/components/shared/ApplicationBoard.tsx]: thx done.

File: packages/mask/src/components/shared/ApplicationBoard.tsx [hunk 13]
```diff
@@ -86,350 +21,27 @@ const useStyles = makeStyles()((theme) => {
...
-                        ) : null,
-                )}
+                {snsAdaptorPlugins
+                    .reduce<Plugin.SNSAdaptor.ApplicationEntry[]>((acc, cur) => {
+                        if (!cur.ApplicationEntries) return acc
+                        return acc.concat(cur.ApplicationEntries ?? [])
+                    }, [])
+                    .sort((a, b) => a.defaultSortingPriority - b.defaultSortingPriority)
+                    .map((X, i) => (
+                        <Fragment key={i}>
```
Comment (Jack-Works) [inline packages/mask/src/components/shared/ApplicationBoard.tsx]: I would prefer to use `PluginID + index` as the key.

**Review by Jack-Works — DISMISSED** [review]

File: packages/mask/src/components/shared/ApplicationBoard.tsx (hunk 13 above)
Comment (hanszhoou) [reply packages/mask/src/components/shared/ApplicationBoard.tsx]: done

**Review by guanbinrui — DISMISSED** [review]

**Comment by guanbinrui** [comment]
@DimensionDev/qa

**Review by Jack-Works — APPROVED** [review]

**Review by guanbinrui — APPROVED** [review]
